src/mlpack/methods/ann/layer/mean_pooling_impl.hpp

```diff
@@ -0,0 +1,201 @@
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_MEAN_POOLING_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_MEAN_POOLING_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "mean_pooling.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename MatType>
+MeanPoolingType<MatType>::MeanPoolingType() :
+    Layer<MatType>()
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+MeanPoolingType<MatType>::MeanPoolingType(
+    const size_t kernelWidth,
+    const size_t kernelHeight,
+    const size_t strideWidth,
+    const size_t strideHeight,
+    const bool floor) :
+    Layer<MatType>(),
+    kernelWidth(kernelWidth),
+    kernelHeight(kernelHeight),
+    strideWidth(strideWidth),
+    strideHeight(strideHeight),
+    floor(floor),
+    channels(0)
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+MeanPoolingType<MatType>::MeanPoolingType(
+    const MeanPoolingType& other) :
+    Layer<MatType>(other),
+    kernelWidth(other.kernelWidth),
+    kernelHeight(other.kernelHeight),
+    strideWidth(other.strideWidth),
+    strideHeight(other.strideHeight),
+    floor(other.floor),
+    channels(other.channels),
+    pooling(other.pooling)
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+MeanPoolingType<MatType>::MeanPoolingType(
+    MeanPoolingType&& other) :
+    Layer<MatType>(std::move(other)),
+    kernelWidth(std::move(other.kernelWidth)),
+    kernelHeight(std::move(other.kernelHeight)),
+    strideWidth(std::move(other.strideWidth)),
+    strideHeight(std::move(other.strideHeight)),
+    floor(std::move(other.floor)),
+    channels(std::move(other.channels)),
+    pooling(std::move(other.pooling))
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+MeanPoolingType<MatType>&
+MeanPoolingType<MatType>::operator=(const MeanPoolingType& other)
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(other);
+    kernelWidth = other.kernelWidth;
+    kernelHeight = other.kernelHeight;
+    strideWidth = other.strideWidth;
+    strideHeight = other.strideHeight;
+    floor = other.floor;
+    channels = other.channels;
+    pooling = other.pooling;
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+MeanPoolingType<MatType>&
+MeanPoolingType<MatType>::operator=(MeanPoolingType&& other)
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(std::move(other));
+    kernelWidth = std::move(other.kernelWidth);
+    kernelHeight = std::move(other.kernelHeight);
+    strideWidth = std::move(other.strideWidth);
+    strideHeight = std::move(other.strideHeight);
+    floor = std::move(other.floor);
+    channels = std::move(other.channels);
+    pooling = std::move(other.pooling);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void MeanPoolingType<MatType>::Forward(
+    const MatType& input, MatType& output)
+{
+  arma::Cube<typename MatType::elem_type> inputTemp(
```

Comment:
It would be really great to add some comments to what is going on here.  It may seem obvious to you since you just implemented it---but it will not be obvious to the person who comes here looking to understand three years from now. :+1:  I would suggest adding some general qualitative comments about what is being done here and why we need to do it.  (At the same time, there is no point in a trivial comment like `Create inputTemp`, so don't write that :))

Reply:
Done.